pkg/provisioners/managers/file-storage/provision.go

```diff
@@ -0,0 +1,204 @@
+func (p *Provisioner) buildDesiredNetworkSet() (map[int]unikornv1.Attachment, error) {
+	desiredSet := make(map[int]unikornv1.Attachment, len(p.fileStorage.Spec.Attachments))
+
+	for _, a := range p.fileStorage.Spec.Attachments {
+		if a.SegmentationID == nil {
+			return nil, fmt.Errorf("%w: network %s has nil SegmentationID", ErrInvalidNetwork, a.NetworkID)
+		}
+
+		if a.IPRange == nil {
+			return nil, fmt.Errorf("%w: network %s has nil IPRange", ErrInvalidNetwork, a.NetworkID)
+		}
+
+		desiredSet[*a.SegmentationID] = a
+	}
+
+	return desiredSet, nil
+}
+
+// buildCurrentNetworkSet returns a set of currently attached VLAN IDs.
+func (p *Provisioner) buildCurrentNetworkSet(ctx context.Context, driver types.Driver) (map[int]struct{}, error) {
+	attachments, err := driver.ListAttachments(ctx, p.fileStorage.Labels[coreconstants.ProjectLabel], p.fileStorage.Name)
+	if err = ignoreNotFound(err); err != nil {
+		return nil, err
+	}
+
+	currentSet := make(map[int]struct{}, len(attachments.Items))
+	for _, a := range attachments.Items {
+		currentSet[a.VlanID] = struct{}{}
+	}
+
+	return currentSet, nil
+}
+
+// attachMissingNetworks attaches networks that are desired but not yet attached.
+func (p *Provisioner) attachMissingNetworks(ctx context.Context, cli client.Client, driver types.Driver, desiredSet map[int]unikornv1.Attachment, currentSet map[int]struct{}, reference string) error {
+	log := log.FromContext(ctx)
+
+	for vlan, attachment := range desiredSet {
+		if _, exists := currentSet[vlan]; exists {
+			continue
+		}
+
+		log.V(1).Info("attaching network", "vlan", vlan)
+
+		// Add references to any resources we consume.
+		if err := manager.AddResourceReference(ctx, cli, &unikornv1.Network{}, client.ObjectKey{Namespace: p.fileStorage.Namespace, Name: attachment.NetworkID}, reference); err != nil {
+			return fmt.Errorf("%w: failed to add network references", err)
+		}
+
+		if err := driver.AttachNetwork(ctx, p.fileStorage.Labels[coreconstants.ProjectLabel], p.fileStorage.Name, &attachment); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// detachStaleNetworks detaches networks that are currently attached but no longer desired.
+func (p *Provisioner) detachStaleNetworks(ctx context.Context, cli client.Client, driver types.Driver, desiredSet map[int]unikornv1.Attachment, currentSet map[int]struct{}, reference string) error {
+	log := log.FromContext(ctx)
+
+	for vlan := range currentSet {
+		if _, desired := desiredSet[vlan]; desired {
+			continue
+		}
+
+		log.V(1).Info("detaching network", "vlan", vlan)
+
+		if err := driver.DetachNetwork(ctx, p.fileStorage.Labels[coreconstants.ProjectLabel], p.fileStorage.Name, vlan); err != nil {
+			return err
+		}
+	}
+
+	// Remove references to networks no longer attached. Since the driver (vast) is the source of truth for attachments but does not track network IDs,
+	// we must list all candidate networks and check which are no longer referenced in fileStorage attachments.
```

Comment:
Actually, you have the file storage CR, you can at least add org/project/region labels to the selector.